### PR TITLE
Add Reporting Templates to the Repo - issue(feature, bug) and pr template.

### LINKED ISSUE #2056: Add Reporting Templates to the Repo
**Is your feature request related to a problem? Please describe.**
It can sometimes be difficult to understand the context of an issue(bug/feature request/etc) or a PR summary

**Describe the solution you'd like**
I'd like to add a template to the repo that guides contributors on how to go about clarifying their ideas/contributions.

**Describe alternatives you've considered**

1. A word doc - which is not remotely as visible. 
2. Leave it till review - doesn't cover issues/request, and might slow down the pr approval process.
3. Direct communication - doesn't scale well.

**Additional context**
This issue was raised as a simple demonstration 😄

### PR DESCRIPTION
**Pull request by obasekiosa**
Closes #2056

## Changes

- Adds auto template for but reporting
- Adds auto template for feature request. 
- Adds auto template for pull request documenting.

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [x] Other: Adds templates to improve contribution flow.

#### Requires testing

- [ ] Yes
- [x] No

#### If yes, did you write tests?

- [ ] Yes
- [x] No

#### Notes on testing
None

#### Requires documentation update

- [ ] Yes
- [x] No

#### Requires explanation in Release Notes

- [ ] Yes
- [x] No

## Remarks
Provides no change to functional code, simply an enhancement in contribution process.

### CODE REVIEW

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+- [ ] No
+
+_If yes, link the PR to the docs update or the issue with the details labeled `docs`. Remove if not applicable._
+
+#### Requires explanation in Release Notes
+
+- [ ] Yes
+- [ ] No
+
+_If yes, fill in the details here. Remove if not applicable._
```
Comment (AnkushinDaniil) [inline .github/pull_request_template.md:52]: I think it's possible to simplify this
```suggestion
## Testing

- [ ] Tests Required
- [ ] Tests Written (if required)

## Documentation

- [ ] Documentation Update Required
- [ ] Include in Release Notes
```
What do you think?

Comment (obasekiosa) [reply .github/pull_request_template.md:52]: I understand the point for further simplification.

but those questions exist to ensure the contributor is thinking about the content of their contribution in relation to what we typical require from a contribution (e.g tests, docs, updating affected areas etc).

Oversimplification usually doesn't help when writing instructions.

The question and answer format, induces thought, reason being the contributor might not necessarily be internal, seasoned to open source contribution or even familiar with software engineering.

**Review by AnkushinDaniil — APPROVED** [review]

**Comment by wojciechos** [comment]
We initially had a PR template, but we decided to remove it because it became a bit of a hassle. See here: https://github.com/NethermindEth/juno/pull/612

**Comment by obasekiosa** [comment]
> We initially had a PR template, but we decided to remove it because it became a bit of a hassle. Tbh, I don't have a strong opinion on it. See here: #612

I took a look at it seems the description section might have been the hassle.

now unassuming i'd ask @wojciechos.  what was the hassle?

**Comment by rianhughes** [comment]
I don't feel strongly either way, no harm in adding it, but we don't seem to get a ton of feature requests either so might be unneeded. I'll leave it up to @kirugan and/or @IronGauntlets :)

**Comment by derrix060** [comment]
I do find annoying having to go through a template to create a PR.  I don't see much harm in not having a template, as long as people write good commit messages, and PR description.

**Comment by derrix060** [comment]
I'd be curious to know what problem exactly you are trying to fix by adding templates on issues and PRs

**Review by derrix060 — CHANGES_REQUESTED** [review]
I personally don't think that it's a good idea to have templates for reporting issues or to create a PR without knowing what is the problem currently.
I can see how it could be important for companies that have really strict guidelines and rules and complience stuff they need to follow, but I don't think we are in this position.


I'm mainly marking as "request changes" because I see that even if we go to the route of having templates, we need to capture why the pr exist, what's the problem it's trying to fix, and I can see that this template doesn't capture this.
I read the commits, the issue and the PR description and I only saw about "what the change is" or "what is done" but not the very important "why"

**Comment by obasekiosa** [comment]
> I do find annoying having to go through a template to create a PR. I don't see much harm in not having a template, as long as people write good commit messages, and PR description.

True it does get annoying. Especially when the list contains things that could have been in the commit messages.

The thing is most commit messages could appear vague to another reader.
Second yea we may even do a way with the list of changes or even the PR templates.

but for the issues templates, its really needed, its hard to understand/replicate an issue without enough context,
Expected behaviour vs current behaviour, where it was ran etc.


Documentation in general can be tasking but its for the sake of the person on the on the other end,
Original poster might be unavailable at the point of addressing, etc.

>I'd be curious to know what problem exactly you are trying to fix by adding templates on issues and PRs

Making issues clearer to understand and addressable even in the absence of OP.

>I personally don't think that it's a good idea to have templates for reporting issues or to create a PR without knowing...

These templates aren't strict, it's extremely loose and they induce clear/precise language/statements when specifying the problem/solution and the ripple effects it could have to other things, like docs.

The idea is the "whys" are addressed in the issues, and the PR's contain the "whats".
In the cases of no issues, remarks covers that.

If we need those sections I could just added them.

**Comment by rianhughes** [comment]
Oh, I thought this template would be for external uses who want to create a feature requests / PR, not Juno devs. 

I don't think we should have to create and fill out a template doc anytime we want to create a PR, it just creates unnecessary overhead, and doesn't solve any problem that we are currently facing. Also, I find the template just gets in the way of understanding what the PR is about.

**Comment by obasekiosa** [comment]
closing this.

**Pull request closed without merging**